util/log: all internal scope calls now ignore -show-logs

BramGruneir opened the pull request:
All of these tests require actual files to be written to disk and when testing with -show-logs it can corrupt the tests.

BramGruneir committed:
util/log, server: all internal scope calls now ignore -show-logs

All of these tests require actual files to be written to disk and when testing with -show-logs it can corrupt the tests.
It also removes the default test directory that was created in main_test.go. That file is no longer needed.

This also fixes testStatusLocalLogs, where the failure was caused by overlapping timestamps.

Fixes #12750
Fixes #14147
Fixes #14014

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14811)
<!-- Reviewable:end -->

tamird commented:
How come these tests don't fail in CI today?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhTR8HyEetFEXmtV1e1:b-oiurmc)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Does CI add -show-logs?

They do still get a logs directory from main_test, but they all use the exact same directory.  I'm guessing that some other bug in windows is leaving some trace in the directory so I get sporadic failures there.  And I got even worse if -show-logs is true.  

On windows after this change, I only get a single failure in util/logs with TestGC.  So I'm looking into that one now.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhTRKF6KrB-TjIS-Wlh:bartxsf)*
<!-- Sent from Reviewable.io -->

tamird commented:
On Tue, Apr 11, 2017 at 3:22 PM, Bram Gruneir <notifications@github.com>
wrote:

> I'm guessing that some other bug in windows is leaving some trace in the
> directory so I get sporadic failures there.


Sounds like something to track down, not paper over.

BramGruneir commented:
Agreed, and I'm working on testgc which I think is the culprit.   But these should be run in isolation.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhTeSF8bbENUOIID4TH:boiz2ab)*
<!-- Sent from Reviewable.io -->

tamird commented:
I really don't understand what you're saying, but I'll let @knz weigh in.

On Tue, Apr 11, 2017 at 4:17 PM, Bram Gruneir <notifications@github.com>
wrote:

> Agreed, and I'm working on testgc which I think is the culprit. But these
> should be run in isolation.
> ------------------------------
>
> Review status: 0 of 3 files reviewed at latest revision, all discussions
> resolved, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhTeSF8bbENUOIID4TH:boiz2ab>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14811#issuecomment-293387635>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPEVJU_M-VGLLeag1ObCmcP3_OLK8ks5ru9_1gaJpZM4M6hrT>
> .
>

knz commented:
@tamird: the tests run fine when you do *not* pass `-show-logs`, because then logs go to files as expected. Indeed theses tests are likely to be (partially) broken with `-show-logs` because that disables file logging for any test that doesn't otherwise enable file logging explicitly (using `enableLogFileOutput`). TC doesn't use `-show-logs` and that's why we didn't notice anything wrong.  

The changes in this PR are thus OK; however the following comment worries me:

> They do still get a logs directory from main_test, but they all use the exact same directory. 

I believe this is wrong, but perhaps I am not understanding this properly. @BramGruneir could you elaborate?

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/test_log_scope.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWETvel24cCG0JaFVI:-KhWETvel24cCG0JaFVJ:b-jw20so) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L51)):*
> ```Go
> }
> 
> var showLogsFlag, showLogs bool
> ```

Please explain here in a comment why you need two separate variables. (I also do not understand it)

---

*[pkg/util/log/test_log_scope.go, line 77 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWEdf-w2LK3L9oVCLW:-KhWEdf-w2LK3L9oVCLX:btkf5yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L77)):*
> ```Go
> // that require the logs go to files.
> func scopeWithoutShowLogs(t tShim, testName string) *TestLogScope {
> 	if testName == "" {
> ```

Factor the determination of `testName` into its own function and reuse it. DRY

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhWDJsljyD4uSp9Z-aj:b-hc8pdu)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
@knz, in main_test, in init, a new log directory is initiated.  If you set -show-logs (before this change) all tests use this log directory instead.  See https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/main_test.go#L35.

When -show-logs is on, since enableLogFileOutput is never set, which is really just setting stderrThreshold to something other than 0.  It will always log to file, since all severities are greater than 0.  See https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/clog.go#L729

@tamird: TestGC is fixed with #14831, but on both windows and osx, running the full test suite in util/log with -show-logs (without this change) will cause failures.  Specifically with the updated TestGC since this other log directory isn't always empty from previous logs files, so the initial check for a single log file finds a second old one from (from TestRollover or TestGetLogReader).  It only fails sometimes since it depends on when the last time the gc logger was run and if it cleaned up the older file or not during the setup for the test.  

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/util/log/test_log_scope.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWETvel24cCG0JaFVI:-KhYHJa4Ji2IXXF1gogt:b-a7kuxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please explain here in a comment why you need two separate variables. (I also do not understand it)
</blockquote></details>

 Since showLogs was being directly set by the flag, it was really ugly to disable it on a per test basis.  By setting the global with each call to Scope, it gets around this.  Another way would be to add a context and only add the flag to the context when calling Scope().  I tried this, but then it means adding contexts all over the place and requiring all functions to use this scope, so an ugly solution to the problem.

I just settled on adding yet another flag to logging.  This should make more sense.  It only gets set in calls to Scope and Close, but is used internally in sb.rotateFile.

I think it would be worthwhile to audit all of the testing flags here and put them all under a mutex.  Maybe a RW one so logging isn't impeded when not testing.

---

*[pkg/util/log/test_log_scope.go, line 77 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWEdf-w2LK3L9oVCLW:-KhYH1mC_hdw0OKdV12K:b-y1jwl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L77)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Factor the determination of `testName` into its own function and reuse it. DRY
</blockquote></details>

I wanted to, but due to the nature of having to lookup different amounts in the stack trace means having to alter the signature of scopeWithShowLogs and it didn't make it any nicer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhVEd3ewib2jglo6U0g:b2x8ovm)*
<!-- Sent from Reviewable.io -->

knz commented:
I propose that you modify `main_test.go` to remove the initialization of `logDir` currently in there. The logging directory should be in full control of `Scope` and a test's default state should be to log to stderr.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/clog.go, line 604 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhbF8dMZ5-kEnoD-TXs:-KhbF8dMZ5-kEnoD-TXt:bobb4wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e02399c045a86b163a866003cef06b3beaef638/pkg/util/log/clog.go#L604)):*
> ```Go
> 
> 	// Flag used for testing, Handled atomically.
> 	showLogs int32
> ```

This needs not be stored in the `loggingT` struct. 

---

*[pkg/util/log/test_log_scope.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWETvel24cCG0JaFVI:-KhbF1g23x7virV0gQzH:b-twbpqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L51)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

 Since showLogs was being directly set by the flag, it was really ugly to disable it on a per test basis.  By setting the global with each call to Scope, it gets around this.  Another way would be to add a context and only add the flag to the context when calling Scope().  I tried this, but then it means adding contexts all over the place and requiring all functions to use this scope, so an ugly solution to the problem.

I just settled on adding yet another flag to logging.  This should make more sense.  It only gets set in calls to Scope and Close, but is used internally in sb.rotateFile.

I think it would be worthwhile to audit all of the testing flags here and put them all under a mutex.  Maybe a RW one so logging isn't impeded when not testing.
</blockquote></details>

I do not think that the `showLogs` variable needs to be protected by a mutex. It doesn't need to live in the `logging` object. There are no concurrent accesses to it.

---

*[pkg/util/log/test_log_scope.go, line 77 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWEdf-w2LK3L9oVCLW:-KhbEUdI1tUGluq9JDZg:btjuh6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L77)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I wanted to, but due to the nature of having to lookup different amounts in the stack trace means having to alter the signature of scopeWithShowLogs and it didn't make it any nicer.
</blockquote></details>

pass the result of `caller.Lookup(1)` as argument to the separate function. Then it can be separate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhbFOeyQViH_dzq8NJk:b-ihdf8a)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I think that's a good idea and I'll do that in a follow up PR.  I'm worried there might be some unknown side effects that I'd rather test for in isolation.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/util/log/clog.go, line 604 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhbF8dMZ5-kEnoD-TXs:-KhbpoZhhMTqZDCaSPiZ:b-mrp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e02399c045a86b163a866003cef06b3beaef638/pkg/util/log/clog.go#L604)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs not be stored in the `loggingT` struct. 
</blockquote></details>

see previous comment.

---

*[pkg/util/log/test_log_scope.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWETvel24cCG0JaFVI:-KhbnAEKt2Yj3nhZwH8C:b-vpl10z) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I do not think that the `showLogs` variable needs to be protected by a mutex. It doesn't need to live in the `logging` object. There are no concurrent accesses to it.
</blockquote></details>

There is use of it within rotate file.  Which is why it needs to be protected.

---

*[pkg/util/log/test_log_scope.go, line 77 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhWEdf-w2LK3L9oVCLW:-KhbnME0JH7gDDfBSHZm:bmfxq1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/64f8c5cfd9c55e39f20647d830de643f2cf9f5f4/pkg/util/log/test_log_scope.go#L77)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

pass the result of `caller.Lookup(1)` as argument to the separate function. Then it can be separate.
</blockquote></details>

Done.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhbmMIqHt8e2u8TtLE1:b-zashnm)*
<!-- Sent from Reviewable.io -->

tamird commented:
@bramgruneir isn't @knz's suggestion a different solution to this same problem? Why would that be follow-up work?

---

Reviewed 1 of 4 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhbqjeXBoFJyQTtnZE-:bth74b9)*
<!-- Sent from Reviewable.io -->

knz commented:
I would rather make showLogs used only by test code, and that it does not appear in the rotate function used in production.

We could sit together and do it together if you'd like/prefer that.

---

Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-Khbs7V3MnTBzS9X3xPS:ba53vvo)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
@tamird, what are are you saying?  Just removing that log directory does not solve the problem, it causes other ones.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhbtdFr-w-B7lLOEJzu:bp9t4m9)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
A number of new chagnes in this PR.  PTAL.  

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhcGBqqgQjLD76m2OOG:bdq6dgf)*
<!-- Sent from Reviewable.io -->

tamird commented:
> This also fixes testStatusLocalLogs, where the failure was caused by overlapping timestamps.

Please elaborate; as I understand, the culprit here is the log files' timestamp resolution.

> Fixes #12750 #14147 #14014

This doesn't work. You need a separate "Fixes" for each one.

```
Fixes #12750.
Fixes #14147.
Fixes #14014.
```

---

Reviewed 3 of 5 files at r3.
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcH5HCX9oZMAjzTxiW:b-cur3dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
> ```Go
> 	// Log an error which we expect to show up on every log file.
> 	timestamp := timeutil.Now().UnixNano()
> 	time.Sleep(ts.Cfg.MaxOffset)
> ```

we need some comments here. also, what does this have to do with maxoffset? if I understood you correctly in person, this should sleep the resolution of the log file name's timestamp, which has nothing at all to do with MaxOffset.

it should go without saying that this needs some comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhcHV58t08GpWfBHaUu:b3ufzyl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/test_log_scope.go, line 53 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcKbJ8gvMgkChXckvQ:-KhcKbJ8gvMgkChXckvR:bxlw8oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/util/log/test_log_scope.go#L53)):*
> ```Go
> var showLogs bool
> 
> func createTestNameFromFuncName(fun string) string {
> ```

quick heads up that I've sent a PR that obviates the need for all this and will likely produce a minor merge conflict (https://github.com/cockroachdb/cockroach/pull/14902).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Fixed the commit message.

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcK7Y-RrPaB84LJtnl:bjtvuuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

we need some comments here. also, what does this have to do with maxoffset? if I understood you correctly in person, this should sleep the resolution of the log file name's timestamp, which has nothing at all to do with MaxOffset.

it should go without saying that this needs some comments.
</blockquote></details>

You're correct, I've adjusted it to be 1 ms.  There was no need for it to be maxoffset, it was just a convenient value to test with.

---

*[pkg/util/log/test_log_scope.go, line 53 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcKbJ8gvMgkChXckvQ:-KhcLbDNyvo08z2JTvMh:bpfepr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/util/log/test_log_scope.go#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

quick heads up that I've sent a PR that obviates the need for all this and will likely produce a minor merge conflict (https://github.com/cockroachdb/cockroach/pull/14902).
</blockquote></details>

Nice.  Much cleaner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhcLXNzHwO8xpuyPiHD:b-9e2k28)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcO9mS-OhZMk8WpDWf:bk4kfy2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

You're correct, I've adjusted it to be 1 ms.  There was no need for it to be maxoffset, it was just a convenient value to test with.
</blockquote></details>

Wait, why is 1ms sufficient? Looks like our filenames only have 1s resolution https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/file.go#L166

On the other hand, individual log lines have a resolution of...10 microseconds, I think?
 https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/clog.go#L437

Please be precise here and reference the specific value that actually matters and where it comes from.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

knz commented:
Probably needs to be rebased / tweaked on top of #14902. Nicely done otherwise.

---

Reviewed 2 of 5 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcT_l9i9OsfQH7iTGU:bo2c05h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Wait, why is 1ms sufficient? Looks like our filenames only have 1s resolution https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/file.go#L166

On the other hand, individual log lines have a resolution of...10 microseconds, I think?
 https://github.com/cockroachdb/cockroach/blob/master/pkg/util/log/clog.go#L437

Please be precise here and reference the specific value that actually matters and where it comes from.
</blockquote></details>

So let me explain the reasoning.
There were two issues
1) there needs to be *some* time between the sampling of the timestamp and the logging event, so that the timestamps are guaranteed to be sufficiently different. How much? We support microsecond precision, so the delay needs to be greater than 1µs; however
2) all this code is using `time.Now()` and is thus subject to time jumps. We know at least on our TC agents that this is something that ahppens. So in order to prevent the test from being flaky we need some delay which is greater than the usual amount of time jumps we're already tolerating otherwise. I suggested MaxOffset since we are already calibrating this to be large enough for the type of time jumps we observe in test clusters.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhcU6gVCM5RUsRKWjez:b-3bpb5t)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcZXYnSOM3Fa8w904o:b-5ld9r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

So let me explain the reasoning.
There were two issues
1) there needs to be *some* time between the sampling of the timestamp and the logging event, so that the timestamps are guaranteed to be sufficiently different. How much? We support microsecond precision, so the delay needs to be greater than 1µs; however
2) all this code is using `time.Now()` and is thus subject to time jumps. We know at least on our TC agents that this is something that ahppens. So in order to prevent the test from being flaky we need some delay which is greater than the usual amount of time jumps we're already tolerating otherwise. I suggested MaxOffset since we are already calibrating this to be large enough for the type of time jumps we observe in test clusters.
</blockquote></details>

I see. Is there any evidence that this test has been plagued by time-jump flakiness? It seems to me that it's very flaky even locally, and that doesn't have anything to do with time jumps.

In any case, this all needs to be included in a comment, whatever the end implementation does (and for the moment I would prefer that we avoid fixing the time jump issue which we haven't seen be a problem yet).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-Khc_eazIqXnWn_umFTN:b-niu9fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I see. Is there any evidence that this test has been plagued by time-jump flakiness? It seems to me that it's very flaky even locally, and that doesn't have anything to do with time jumps.

In any case, this all needs to be included in a comment, whatever the end implementation does (and for the moment I would prefer that we avoid fixing the time jump issue which we haven't seen be a problem yet).
</blockquote></details>

+1 on making a comment.

Then for the delay I simply recommend a loop that waits until time.Now() is greated by 1 microsecond, e.g. `for time.Now().Sub(prevTS) < time.Microsecond {}` or something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcbHEfVjYcxV3koxqz:bmuovp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

+1 on making a comment.

Then for the delay I simply recommend a loop that waits until time.Now() is greated by 1 microsecond, e.g. `for time.Now().Sub(prevTS) < time.Microsecond {}` or something.
</blockquote></details>

How is that any different better than just sleeping for 1 microsecond?  (and as Tamir mentioned earlier, it would have to be 10 microseconds)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-Khcbh4VICjr7eFD8xig:b-tbhquz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

How is that any different better than just sleeping for 1 microsecond?  (and as Tamir mentioned earlier, it would have to be 10 microseconds)
</blockquote></details>

Depending on how Sleep() is implemented, sleeping for the exact period you need might sleep for a duration a fraction smaller than that, and thus insufficient for the test.
2 times that period would be probably sufficient, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcjRuyTCog-U1cs-bm:b-9ajiv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Depending on how Sleep() is implemented, sleeping for the exact period you need might sleep for a duration a fraction smaller than that, and thus insufficient for the test.
2 times that period would be probably sufficient, though.
</blockquote></details>

I think sleeping for the desired duration is fine for now, and you can validate that under stress.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-KhcjrCvzv0NddFOrZwo:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think sleeping for the desired duration is fine for now, and you can validate that under stress.
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Rebased and good to go.  I now await a LGTM and a green build.

---

Review status: 1 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/server/status_test.go, line 201 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcH5HCX9oZMAjzTxiV:-Khct6P83UvszDKbBrcb:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c4d8cd69d578e79cd755f0b87baa5bf9ad34e24/pkg/server/status_test.go#L201)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Done!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-Khct82rR-BlARFjYF8x:b-3t3yj3)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Just noticed a wee bit of merge skew, fixing.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhctWxVzCsfuxgDYz1q:b8oi6ls)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Fixed.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhcuxB0aW0q3yEf4RIA:b-lx4fjp)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod clog_test scopes

---

Reviewed 3 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/server/status_test.go, line 200 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcxtSWzElB5HE620aV:-KhcxtSXv7gnmU4Ir6XE:b-ygs2a1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/server/status_test.go#L200)):*
> are

is

---

*[pkg/server/status_test.go, line 206 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcxxsDhIAbtFo1-z_j:-KhcxxsDhIAbtFo1-z_k:bc9cpxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/server/status_test.go#L206)):*
> ```Go
> 	// the log entires we're looking for. The value of 20 μs was chosen because
> 	// the log timestamps have a fidelity of 10 μs and thus doubling that should
> 	// be a sufficient buffer.
> ```

please link to the source of truth on this, and extract a constant to avoid repeating the value N times.

---

*[pkg/util/log/clog_test.go, line 104 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-Khcyyd1ju5EMZNJG0Y-:-Khcyyd1ju5EMZNJG0Y0:b-56qtlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/util/log/clog_test.go#L104)):*
> ```Go
> // Test that Info works as advertised.
> func TestInfo(t *testing.T) {
> 	s := ScopeWithoutShowLogs(t)
> ```

@knz should sign off on this file, I don't understand the implication of adding these scopes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-Khcz4av8UWNWZTnhzOt:b8e4q9o)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 4 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/server/status_test.go, line 200 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcxtSWzElB5HE620aV:-Khd5PcLrclAkAiWWxbd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/server/status_test.go#L200)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> are

is
</blockquote></details>

Done.

---

*[pkg/server/status_test.go, line 206 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-KhcxxsDhIAbtFo1-z_j:-Khd5lQpApGZYV_IrhQk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/server/status_test.go#L206)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

please link to the source of truth on this, and extract a constant to avoid repeating the value N times.
</blockquote></details>

Done.

---

*[pkg/util/log/clog_test.go, line 104 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-Khcyyd1ju5EMZNJG0Y-:-Khd6DF3UcV4zxLWAHzd:bj14y45) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/util/log/clog_test.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

@knz should sign off on this file, I don't understand the implication of adding these scopes.
</blockquote></details>

They all call contents(), which looks at the buffer, and that buffer only gets populated without the -show-logs flag.  But I'll wait for his sign off before merging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/util/log/clog_test.go, line 104 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-Khcyyd1ju5EMZNJG0Y-:-KhfUHNahBzPeV7_BxmK:b-6d3w0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb131f6b4ff320dbb38421769032d90fa990804/pkg/util/log/clog_test.go#L104)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

They all call contents(), which looks at the buffer, and that buffer only gets populated without the -show-logs flag.  But I'll wait for his sign off before merging.
</blockquote></details>

The change is correct!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14811#-:-KhfUPuSjVoavB7UyCUG:bnfp4nl)*
<!-- Sent from Reviewable.io -->
